### PR TITLE
🐛 Ensure all GitHub releases are fetched when searching provider versions

### LINKED ISSUE #8233: clusterctl only considers 30 github releases
When clusterctl looks for releases for providers it looks up the https://api.github.com/repos/kubernetes-sigs/cluster-api/releases endpoint. This endpoint by default only lists the most recent 30 releases. Even if an earlier release does exist clusterctl will not find it if it's not in the most recent 30 releases.

The current impact of this is that clusterctl v0.4.X and clusterctl v0.3.X no longer work to initialize providers. A workaround is to pull the provider yamls manually and put them in a local clusterctl repository.

v0.4.X and v0.3.X of Cluster API are out of date so there is no need to update those older versions to fix this bug. This should be fixed in supported versions of Cluster API as we don't control the number of releases done by providers and different approaches e.g. many published pre-releases, could result in an older supported version not being installable. 

The API call should be able to look through subsequent pages of results in order to find older valid tags. Note: This should also be checked for the gitlab API to see if it suffers from a similar limitation. 

/kind bug
/area clusterctl

### PR DESCRIPTION
**Pull request by abhay-krishna**
**What this PR does / why we need it**:
This PR improves the logic of fetching GitHub releases when looking for versions in provider repositories. The `ListReleases` API call is paginated and returns only the latest 30 releases. So we need to issue multiple calls in series to fetch all releases for a repository.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8233 (**NOTE:** This PR handles only GitHub, it doesn't include a GitLab fix.)

### CODE REVIEW

File: cmd/clusterctl/client/repository/repository_github.go
Comment (CecileRobertMichon) [inline cmd/clusterctl/client/repository/repository_github.go:288]: is there any unit test that can be added to test the bug this is fixing?

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:288]: I did look into the test file for this source code, and it seems like it's using a fake request handler. Would you like me to add a test case for this there?

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:288]: The source code itself has been validated locally [here](https://github.com/kubernetes-sigs/cluster-api/issues/8233#issuecomment-1456989355).

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:288]: I have modified the getVersions Test method to construct a fake releases listing and emulate three different scenarios, one where an valid page number is given, one where an invalid page number is given and one which fetches all the (fake) releases. This is essentially mocking the ListReleases API call.

File: cmd/clusterctl/client/repository/repository_github.go
Comment (mboukhalfa) [inline cmd/clusterctl/client/repository/repository_github.go:314]:
```suggestion
	for response.NextPage != 0 {
				releases, response, listReleasesErr = client.Repositories.ListReleases(context.TODO(), g.owner, g.repository, &github.ListOptions{Page:  response.NextPage})
				if listReleasesErr != nil {
					retryError = g.handleGithubErr(listReleasesErr, "failed to get the list of releases")
					// return immediately if we are rate limited
					if _, ok := listReleasesErr.(*github.RateLimitError); ok {
						return false, retryError
					}
					return false, nil
				}
				allReleases = append(allReleases, releases...)
			}
```

File: cmd/clusterctl/client/repository/repository_github.go
Comment (chrischdi) [inline cmd/clusterctl/client/repository/repository_github.go:313]: This lines seem to be almost the same as 208 to 297.

Could we maybe combine these in a single for loop?

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:313]: I had them duplicated to apply the rate limiting to each call to the ListReleases API.

File: cmd/clusterctl/client/repository/repository_github.go
Comment (jackfrancis) [inline cmd/clusterctl/client/repository/repository_github.go:421]: nit: because this is not generalized we might want to rename the `retryError` return var name to just `err`

Comment (jackfrancis) [reply cmd/clusterctl/client/repository/repository_github.go:421]: I see we're using `err` for the func input error object, so ignore that for the time being.

However, another point: this new `checkGithubErr` on top of `handleGithubErr` doesn't fully make sense to me, as we're testing if the `err` type is `github.RateLimitError` twice now. Is that intended?

Comment (jackfrancis) [reply cmd/clusterctl/client/repository/repository_github.go:421]: I think the entire usage surface area of `handleGithubErr` can be refactored, how about we revert all of these changes in this PR so we can simply focus on the improvements to pagination, and then one of us can do that refactor work in a follow-up PR? (happy to do it myself)

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:421]: basically it looks like `handleGithubErr` returns two types of errors, one for the rate limit and one for any other reason which might have caused the error.
```go
func (g *gitHubRepository) handleGithubErr(err error, message string, args ...interface{}) error {
	if _, ok := err.(*github.RateLimitError); ok {
		return errors.New("rate limit for github api has been reached. Please wait one hour or get a personal API token and assign it to the GITHUB_TOKEN environment variable")
	}
	return errors.Wrapf(err, message, args...)
}
```
But this check
```
retryError := g.handleGithubErr(err, message, args...)
// return immediately if we are rate limited
if _, ok := listReleasesErr.(*github.RateLimitError); ok {
	return false, retryError
}
return false, nil
```
 indicates that we are only surfacing the first error to the PollImmediate method, because that indicates a rate limit and we should stop the retry.
```
PollImmediate tries a condition func until it returns true, an error, or the timeout is reached.
```

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:421]: I guess we can return nil from `handleGithubErr` in the second case, and then we needn't check again for `github.RateLimitError` twice since it will only ever return error on rate limit

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:421]: It seems like an antipattern to mask an error as nil, but that's what seem to be doing [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/client/repository/repository_github.go#L294)

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:421]: Sure, I can revert the changes

**Comment by abhay-krishna** [comment]
Thanks for the lgtm @jackfrancis! Is there a code freeze now or is this good to be merged?

**Review by killianmuldoon — COMMENTED** [review]
Mostly just nits, but I don't think the unit test here really gives additional coverage after stepping through it. It's fine here not to mock the external API so IMO dropping that test / reverting it to the way it was before is okay.

I've tested this locally though and this collects ~120 releases of CAPI, letting it pick up otherwise "lost" releases - e.g. v0.4.8

Thanks for working on this @abhay-krishna!

**Comment by abhay-krishna** [comment]
> Mostly just nits, but I don't think the unit test here really gives additional coverage after stepping through it. It's fine here not to mock the external API so IMO dropping that test / reverting it to the way it was before is okay.
> 
> I've tested this locally though and this collects ~120 releases of CAPI, letting it pick up otherwise "lost" releases - e.g. v0.4.8
> 
> Thanks for working on this @abhay-krishna!

Yeah that was my thinking too, because at that point we really aren't mocking the CAPI method, but the upstream ListReleases API.

File: cmd/clusterctl/client/repository/repository_github.go
Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go]: Done!

**Comment by abhay-krishna** [comment]
@killianmuldoon Thanks for reviewing, I fixed all the comments and superfluous newlines. I would still like to keep the `getVersions` unit test to mock getting release versions from the paginated GH API.

**Comment by killianmuldoon** [comment]
Thanks for the updates! The issue with the unit test as-is is that it doesn't actually add any coverage for the new code - there's no response that contains "NextPage" so this is never hit by the testing code AFAIU.

**Comment by abhay-krishna** [comment]
> Thanks for the updates! The issue with the unit test as-is is that it doesn't actually add any coverage for the new code - there's no response that contains "NextPage" so this is never hit by the testing code AFAIU.

Makes sense, I have reverted the changes to the test.

**Review by ykakarap — CHANGES_REQUESTED** [review]
I see that there was already a discussion on test. Looks like in the current state of the PR the logic that loops through all pages to get all the releases is not actually tested. 
I think we should test the case where the list releases call returns a response with `NextPage != 0` and that our `getVersions` call actually lists all the releases from multiple pages.
@killianmuldoon WDYT?

**Comment by abhay-krishna** [comment]
> I see that there was already a discussion on test. Looks like in the current state of the PR the logic that loops through all pages to get all the releases is not actually tested. I think we should test the case where the list releases call returns a response with `NextPage != 0` and that our `getVersions` call actually lists all the releases from multiple pages. @killianmuldoon WDYT?

However, the NextPage value is not an input Parameter to getVersions, but rather to ListReleases right? And we can't make calls to the actual GitHub API with the fake client. In my previous version of the test (which I have now reverted), I had the test loop through a large number of releases with a small page size, so as to emulate the next page workflow. The mocked method doesn't return a NextPage but returns a set of results from the current page number, which means the previous value of NextPage was a non-zero value.

**Comment by ykakarap** [comment]
> > I see that there was already a discussion on test. Looks like in the current state of the PR the logic that loops through all pages to get all the releases is not actually tested. I think we should test the case where the list releases call returns a response with `NextPage != 0` and that our `getVersions` call actually lists all the releases from multiple pages. @killianmuldoon WDYT?
> 
> However, the NextPage value is not an input Parameter to getVersions, but rather to ListReleases right? And we can't make calls to the actual GitHub API with the fake client. In my previous version of the test (which I have now reverted), I had the test loop through a large number of releases with a small page size, so as to emulate the next page workflow. The mocked listReleases method is called in a loop and returns a set of results from the current page number in each iteration, which emulates the scenario that the previous value of NextPage was a non-zero value.

The FakeGithubClient we use in the tests can be used to mock any github API calls without making the actual github API call.
I dug a little and found that `response.NextPage` is populated based on the headers of the response.
Example header:
```
Link: <https://api.github.com/repositories/124157517/releases?per_page=100&page=2>; rel="next"
```
If the response has the above header it will set `response.NextPage` to `2`.

> I had the test loop through a large number of releases with a small page size, so as to emulate the next page workflow.

I assume you are referring to this commit: [58218ec6bce6badc6fa82458fd0b479ad292e3d1](https://github.com/kubernetes-sigs/cluster-api/commit/58218ec6bce6badc6fa82458fd0b479ad292e3d1#diff-32e56b9113cee0f60cc29a7a50fc75ed96a3b7e3d18171a776867d9c721234bc) (I hope I got the right commit 🙃)
In this case the response does not have `NextPage` so it doesnt not cover the desired case.

I would recommend that we test it by mocking a response that sets `NextPage` and therefore `getVersions()` has to loop through multiple "pages" to build the full list. Example to mock multiple pages:
```go
mux.HandleFunc("/repos/o/r4/releases", func(w http.ResponseWriter, r *http.Request) {
	testMethod(t, r, "GET")
	page := r.URL.Query().Get("page")
	if page == "" {
		// Page 1 of releases
		w.Header().Set("Link", `<https://api.github.com/repositories/12345/releases?page=2>; rel="next"`)
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":1, "tag_name": "v0.4.0"},`)
		fmt.Fprint(w, `{"id":2, "tag_name": "v0.4.1"}`)
		fmt.Fprint(w, `]`)
	} else {
		// Page 2 of releases
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":3, "tag_name": "v0.4.2"},`)
		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-alpha"}`) // Pre-release
		fmt.Fprint(w, `]`)
	}
})
```
With this we can add a test that checks that we get back 4 tags and this should be enough to test that `getVersions` loops over multiple pages to get all the releases.

**Comment by abhay-krishna** [comment]
> > > I see that there was already a discussion on test. Looks like in the current state of the PR the logic that loops through all pages to get all the releases is not actually tested. I think we should test the case where the list releases call returns a response with `NextPage != 0` and that our `getVersions` call actually lists all the releases from multiple pages. @killianmuldoon WDYT?
> > 
> > 
> > However, the NextPage value is not an input Parameter to getVersions, but rather to ListReleases right? And we can't make calls to the actual GitHub API with the fake client. In my previous version of the test (which I have now reverted), I had the test loop through a large number of releases with a small page size, so as to emulate the next page workflow. The mocked listReleases method is called in a loop and returns a set of results from the current page number in each iteration, which emulates the scenario that the previous value of NextPage was a non-zero value.
> 
> The FakeGithubClient we use in the tests can be used to mock any github API calls without making the actual github API call. I dug a little and found that `response.NextPage` is populated based on the headers of the response. Example header:
> 
> ```
> Link: <https://api.github.com/repositories/124157517/releases?per_page=100&page=2>; rel="next"
> ```
> 
> If the response has the above header it will set `response.NextPage` to `2`.
> 
> > I had the test loop through a large number of releases with a small page size, so as to emulate the next page workflow.
> 
> I assume you are referring to this commit: [58218ec6bce6badc6fa82458fd0b479ad292e3d1](https://github.com/kubernetes-sigs/cluster-api/commit/58218ec6bce6badc6fa82458fd0b479ad292e3d1#diff-32e56b9113cee0f60cc29a7a50fc75ed96a3b7e3d18171a776867d9c721234bc) (I hope I got the right commit 🙃) In this case the response does not have `NextPage` so it doesnt not cover the desired case.
> 
> I would recommend that we test it by mocking a response that sets `NextPage` and therefore `getVersions()` has to loop through multiple "pages" to build the full list. Example to mock multiple pages:
> 
> ```go
> mux.HandleFunc("/repos/o/r4/releases", func(w http.ResponseWriter, r *http.Request) {
> 	testMethod(t, r, "GET")
> 	page := r.URL.Query().Get("page")
> 	if page == "" {
> 		// Page 1 of releases
> 		w.Header().Set("Link", `<https://api.github.com/repositories/12345/releases?page=2>; rel="next"`)
> 		fmt.Fprint(w, `[`)
> 		fmt.Fprint(w, `{"id":1, "tag_name": "v0.4.0"},`)
> 		fmt.Fprint(w, `{"id":2, "tag_name": "v0.4.1"}`)
> 		fmt.Fprint(w, `]`)
> 	} else {
> 		// Page 2 of releases
> 		fmt.Fprint(w, `[`)
> 		fmt.Fprint(w, `{"id":3, "tag_name": "v0.4.2"},`)
> 		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-alpha"}`) // Pre-release
> 		fmt.Fprint(w, `]`)
> 	}
> })
> ```
> 
> With this we can add a test that checks that we get back 4 tags and this should be enough to test that `getVersions` loops over multiple pages to get all the releases.

Thanks for the pointers @ykakarap! I have pushed a change to the test which is a generic implementation of your mocking example for any list of releases/page size/number of pages. Hope that should do it!

File: cmd/clusterctl/client/repository/repository_github_test.go
Comment (ykakarap) [inline cmd/clusterctl/client/repository/repository_github_test.go:376]: How about this:
```go
mux.HandleFunc("/repos/o/r1/releases", func(w http.ResponseWriter, r *http.Request) {
	testMethod(t, r, "GET")
	page := r.URL.Query().Get("page")
	switch page {
	case "", "1":
		// Page 1
		w.Header().Set("Link", `<https://api.github.com/repositories/12345/releases?page=2>; rel="next"`) // Link to page 2
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":1, "tag_name": "v0.4.0"},`)
		fmt.Fprint(w, `{"id":2, "tag_name": "v0.4.1"}`)
		fmt.Fprint(w, `]`)
	case "2":
		// Page 2
		w.Header().Set("Link", `<https://api.github.com/repositories/12345/releases?page=3>; rel="next"`) // Link to page 3
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":3, "tag_name": "v0.4.2"},`)
		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-alpha"}`) // prerelease
		fmt.Fprint(w, `]`)
	case "3":
		// Page 3 (last page)
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-beta"},`) // prerelease
		fmt.Fprint(w, `{"id":5, "tag_name": "foo"}`)          // no semantic version tag
		fmt.Fprint(w, `]`)
	default:
		t.Fatalf("unexpected page requested")
	}
}
```
(also adjust the corresponding test's `want`. With this we will also not need `returnPageResults` and can be dropped)

IMO, it makes the tests more readable and easy to maintain.

Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github_test.go]: Thanks! Updated to use a static release listing and pagination. True, there's no need to complicate a unit testing with custom functions.

**Comment by abhay-krishna** [comment]
@ykakarap any changes to be done here?

File: cmd/clusterctl/client/repository/repository_github.go
Comment (ykakarap) [inline cmd/clusterctl/client/repository/repository_github.go:288]: Looks like the option to set `PerPage: 100` was dropped during the rebase. Can you please add it back. We would want to keep the number of API calls as small as possible.

**Comment by ykakarap** [comment]
@abhay-krishna The new test changes look good but looks like the rebase dropped the `PerPage` setting when listing releases. This PR is ready to merge once you add that back. :)

File: cmd/clusterctl/client/repository/repository_github.go
Comment (abhay-krishna) [reply cmd/clusterctl/client/repository/repository_github.go:288]: Ah my bad, adding it back

**Comment by abhay-krishna** [comment]
> 

@ykakarap Thanks for catching that, I've added back the `PerPage` parameter for the ListReleases call. PTAL!

**Comment by ykakarap** [comment]
This looks great. Thanks for working on this 🙂 

/lgtm

**Comment by abhay-krishna** [comment]
> This looks great. Thanks for working on this 🙂 
> 
> /lgtm

Thanks a lot for your review and suggestions!

**Comment by abhay-krishna** [comment]
@sbueringer @ykakarap If I rebase and push, it'll remove the `lgtm` label, and I'll need to 🐛  the maintainers again, so holding off on doing so. There are no conflicts here (yet), so should be a clean merge either ways.

**Comment by ykakarap** [comment]
I don't see any conflicts and looks like all the commits are squashed. Is there a need to rebase?

**Comment by sbueringer** [comment]
/assign @fabriziopandini 
If you have some time, would be great if you can take over the final review/approval

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve

**Comment by sbueringer** [comment]
@CecileRobertMichon @ykakarap  Does it make sense to backport this as a bug fix?

**Comment by jackfrancis** [comment]
@sbueringer You didn't ask for my opinion, but here it is anyways :)

@killianmuldoon said in the issue description:

> This should be fixed in supported versions of Cluster API as we don't control the number of releases done by providers and different approaches e.g. many published pre-releases, could result in an older supported version not being installable.

That suggests a backport to the set of branches that can cut patch releases for "supported versions", right?

**Comment by sbueringer** [comment]
> @sbueringer You didn't ask for my opinion, but here it is anyways :)

Always appreciated :)

**Comment by sbueringer** [comment]
Would be great if someone can create cherry-picks for 1.3 and 1.2

**Comment by ykakarap** [comment]
I can do the release-1.2 cherry-pick.

**Pull request merged**
